### PR TITLE
Triangle terror

### PR DESCRIPTION
**Pull request by bspanoghe**

### CODE REVIEW

**Comment by MichielStock** [comment]
I think the inefficient version of `drawtriangle` should be efficient when you have many triangles you color in batch. So, to draw your image, you only have to do a single pass over all the images.

**Comment by MichielStock** [comment]
- fix functions in camelcase, e.g. `triangleTournament` => `triangletournament`
- keep yourself to one file per piece
- I think you can process all the triangles for an image at once, when you have many triangles, almost all pixels fall within one?
- an individual = a collection of triangles, no?
- `mutateTriangle` should have parameters to tweak the intensity of the mutations.
- allocate a single array to compute your image, this will save a lot of computation time
- remove `safebox.jl`

**Comment by MichielStock** [comment]
put all your test files into the correct format

**Comment by MichielStock** [comment]
and your test does not work anyway

**Comment by MichielStock** [comment]
Good job, but a bit of reorganisation might be in order 🧹 

Might be interesting to [read this](https://en.wikisource.org/wiki/Flatland_(first_edition)) over the holidays.

**Comment by MichielStock** [comment]
Needs to be reviewed tomorrow by @f6v and @kirstvh.

File: Project.toml [hunk 1]
```diff
@@ -4,6 +4,11 @@ authors = ["michielstock <michielfmstock@gmail.com>"]
...
 
 [deps]
+Colors = "5ae59095-9a9b-59fe-a467-6f913c188581"
 Documenter = "e30172f5-a6a5-5a46-863b-614d45cd2de4"
+Images = "916415d5-f1e6-5110-898d-aaa5f9f070e0"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
+PlutoUI = "7f904dfe-b85e-4ff6-b463-dae2292396a8"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```
Comment (f6v) [inline Project.toml]: Seems like `QuartzImageIO` is a dependency as well. Could you add it here, please?

File: docs/src/man/polygon.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
+## polygon
```
Comment (f6v) [inline docs/src/man/polygon.md:1]: Maybe a bit more formal description of the algorithm and some references are in order?

File: src/safebox.jl [hunk 3]
```diff
@@ -0,0 +1,123 @@
+"""
```
Comment (f6v) [inline src/safebox.jl]: That's exactly what Git and GitHub are for. You can create a commit for deleting files and always roll back.

File: test/drawtriangle_test.jl [hunk 4]
```diff
@@ -0,0 +1,58 @@
+using STMOZOO.polygon, Images, Colors
```
Comment (f6v) [inline test/drawtriangle_test.jl]: Do you still need this file here?

File: test/in_test.jl [hunk 5]
```diff
@@ -0,0 +1,22 @@
+using STMOZOO.polygon
```
Comment (f6v) [inline test/in_test.jl]: Do you still need this file? You can always restore it if you make a commit in Git.

File: test/polygon.jl [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    @testset "samesideofline" begin
+        line = (1 + 1im, 2 + 3im)
+        point1 = 0 + 3im #left of line
+        point2 = 1 + 4im #left
+        point3 = 2 + 3.3im #left
+        point4 = 1.5 + 1im #right
+        point5 = 4 + 3im #right
+        point6 = 4 + 4im #right
+
+        # points = [point1, point2, point3, point4, point5, point6]
```
Comment (f6v) [inline test/polygon.jl]: Remove the commented out code, please.

File: test/polygon.jl [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        testpoint2[1] = 4 + 1im #not
+        testpoint2[2] = 3 + 0.5im #not
+        testpoint2[3] = 5 + 3im #not
+        testpoint2[4] = 1 + 2im #not
+        testpoint2[5] = 4 + 2im #yes
+        testpoint2[6] = 3 + 2im #yes
+        testpoint2[7] = 3.5 + 2.5im #yes
+        testpoint2[8] = 3.1 + 3.5im #yes
+
+        for i in 1:4
```
Comment (f6v) [inline test/polygon.jl:61]: It's a bit annoying to keep track of indices. I'd create two arrays:
```
validtestpoints = ...
invalidtestpoints = ...
```
and then something like(pseudocode):
```
for point in validtestponts
   @test point in tringle1
end

for point in invalidtestponts
   @test point is not in tringle1
end

```

File: test/polygon.jl [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+            @test (testpoint1[i] in T1) == false
+            @test (testpoint2[i] in T2) == false
+        end
+        for i in 5:8
+            @test (testpoint1[i] in T1) == true
+            @test (testpoint2[i] in T2) == true
```
Comment (f6v) [inline test/polygon.jl]: I'm not sure, but can't it be just `@test (testpoint1[i] in T1)`?

File: test/polygon.jl [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+    testtriangle[6] = Triangle(446 + 193im, 401 + 246im, 495 + 298im, RGB(1, 0, 0)) # STUPID triangle (all points outside of canvas)
+    testtriangle[7] = Triangle(59 + 189im, 58 + 270im, 57 + 5im, RGB(1, 0, 1)) # STUPID triangle (too thin)
+    testtriangle[8] = Triangle(-59 + 189im, -20 + 63im, -200 + 69im, RGB(1, 0, 1)) # STUPID triangle (points outside of canvas, all x are wrong)
+    testtriangle[9] = Triangle(40 + 150im, 239 + 151im, 20 + 150im, RGB(1, 0, 1)) # STUPID triangle (too thin)
+    testtriangle[10] = Triangle(211 + 209im, 338 + 334im, 544 + 534im, RGB(1, 0, 0)) # STUPID triangle (very elongated)
+
+
+    @testset "checktriangle" begin
+        for i in 1:5
```
Comment (f6v) [inline test/polygon.jl:89]: Same as above

File: test/polygon.jl [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+    @testset "getboundaries" begin
+        @test getboundaries(testtriangle[1], m, n) == (20, n, 30, m)
+        @test getboundaries(testtriangle[5], m, n) == (1, 200, 50, m)
+    end
+
+    @testset "generatetriangle" begin
+        @test typeof(generatetriangle(m, n)) == Triangle
+        anystupids = false
+        for i in 1:100
```
Comment (f6v) [inline test/polygon.jl:103]: Not an issue here, but testing functions with a random component can result in what's called a *flaky test*.  Just FYI.

File: test/polygon.jl [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+            anystupids = anystupids || checktriangle(generatetriangle(m, n), m, n) # Has one of the generated triangles been a stupid triangle?
+        end
+        @test anystupids == false
```
Comment (f6v) [inline test/polygon.jl:106]: You could also make it one-liner:

```
@test !(map(_ -> checktriangle(generatetriangle(m, n), m, n), 1:100) |> any)
```

File: test/polygon.jl [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+        @test colordiffsum(bluebox, purplebox) < colordiffsum(greenbox, purplebox)
+    end
+
+    @testset "triangletournament" begin
+        target = fill(RGB(0, 1, 1), 100, 100)
+        pop = generatepopulation(3, 10, target)
+        sort!(pop, by = x -> x[3])
+        anyweakest = false
+        for i in 1:100
+            anyweakest = anyweakest || triangletournament(pop, 3) == pop[3] # Did the worst individual of the population ever win? (Shouldn't happen)
```
Comment (f6v) [inline test/polygon.jl:126]: same as above

File: src/polygon.jl [hunk 13]
```diff
@@ -0,0 +1,508 @@
...
+
+"""
+A type Shape so the code can be made to work with e.g. rectangles as well. For now, only triangles are supported.
+"""
+abstract type Shape end
+
+"""
+    Triangle <: Shape
+
+I define triangles with the coordinates of their 3 points and a color.
```
Comment (f6v) [inline src/polygon.jl]: Try avoiding comments if they don't add new information. The code for `Triangle` is already crystal clear. But If you adjust the definition, there's absolutely no need for the comment:

```
struct Triangle <: Shape
     point1::ComplexF64 
     point2::ComplexF64
     point3::ComplexF64
     color::RGB{Float64}
 end
```

File: src/polygon.jl [hunk 14]
```diff
@@ -0,0 +1,508 @@
...
+function drawtriangle(triangle::Triangle, img::Array)
+    m, n = size(img)  # the original one needed to allocate memory
+    
+    xmin, xmax, ymin, ymax = getboundaries(triangle, m, n)
+
+    j_left = xmin
+    j_right = xmax
+    left_points = Vector{Int}(undef, 2)
+    right_points = Vector{Int}(undef, 2)
+    counter = 1
```
Comment (f6v) [inline src/polygon.jl]: What is It counting? Is a more descriptive name possible?

File: src/polygon.jl [hunk 15]
```diff
@@ -0,0 +1,508 @@
...
+A stupid triangle is a triangle which:
+- Is much too thin (a couple of pixels thin)
+- Has all 3 points outside of the canvas
+- Is very stretched out
+These stupid triangles can all mess with the drawtriangle function, which is why checking for them and weeding them out is essential.
+Otherwise we get errors and boy do we hate errors.
+"""
+function checktriangle(triangle::Triangle, m::Int, n::Int)
+    # Rather than using the actual boundaries of the canvas, we use a slightly smaller rectangle for safety reasons
+    s = 10 # safety factor
```
Comment (f6v) [inline src/polygon.jl]: Why `10` and not `0.1` or `100`? Is there any rationale?

File: src/polygon.jl [hunk 16]
```diff
@@ -0,0 +1,508 @@
...
+    p3_out_of_canvas = (x3 <= s || x3 >= (n - s)) || (y3 <= s || y3 >= (m - s))
+    pointsAreStupid = all([p1_out_of_canvas, p2_out_of_canvas, p3_out_of_canvas]) && return true # All points are (almost) outside of the canvas, that's no bueno
+
+    # Very elongated triangles cause trouble: check for these as well
+    # Credit for idea of using the centroid: Her Magnificence, Gender Equality Expert Dr. Roets
+    midpoint = (triangle.p1 + triangle.p2 + triangle.p3)/3
+    dist1 = abs(triangle.p1 - midpoint)
+    dist2 = abs(triangle.p2 - midpoint)
+    dist3 = abs(triangle.p3 - midpoint)
+    isStretchyBoi = max(dist1, dist2, dist3) > 2*min(dist1, dist2, dist3) && return true
```
Comment (f6v) [inline src/polygon.jl]: I appreciate your sense of humour, and I think good jokes are absolutely necessary in any team. However, the code should be as unambiguous as possible. Some people might not get what you're talking about(and you have to add comments, which further pollutes the code) or be in a wrong mood. I'd suggest renaming `isStretchyBoi` to `iselongated`.

File: src/polygon.jl [hunk 17]
```diff
@@ -0,0 +1,508 @@
...
+
+    max_x = max(x1, x2, x3)
+    min_x = min(x1, x2, x3)
+    max_delta_x = max_x - min_x
+    XisStupid = max_delta_x <= s && return true # Too thin
+
+    p1_out_of_canvas = (x1 <= s || x1 >= (n - s)) || (y1 <= s || y1 >= (m - s))
+    p2_out_of_canvas = (x2 <= s || x2 >= (n - s)) || (y2 <= s || y2 >= (m - s))
+    p3_out_of_canvas = (x3 <= s || x3 >= (n - s)) || (y3 <= s || y3 >= (m - s))
+    pointsAreStupid = all([p1_out_of_canvas, p2_out_of_canvas, p3_out_of_canvas]) && return true # All points are (almost) outside of the canvas, that's no bueno
```
Comment (f6v) [inline src/polygon.jl:270]: I think these code blocks can be extracted into dedicated functions. `checktriangle` has too many return statements now, it's a bit hard to follow the control flow.

File: src/polygon.jl [hunk 18]
```diff
@@ -0,0 +1,508 @@
...
+end
+
+"""
+    generatetriangle(m::Int, n::Int)
+
+Generates a random triangle (a non-stupid one at that).
+"""
+function generatetriangle(m::Int, n::Int)
+    badTriangle = true
+    t = nothing
```
Comment (f6v) [inline src/polygon.jl]: Please, avoid one-letter variable names, it's code smell.

File: src/polygon.jl [hunk 19]
```diff
@@ -0,0 +1,508 @@
...
+    while badTriangle
+        center = complex(rand()*1.1*n - 0.05*n, rand()*1.1*m - 0.05*m)
```
Comment (f6v) [inline src/polygon.jl]: Where do these constants(`1.1`, `4/5` etc.) come from?

File: src/polygon.jl [hunk 20]
```diff
@@ -0,0 +1,508 @@
...
+
+Picks 2 random triangles from a population and returns the one with the best (= lowest) score.
+"""
+function triangletournament(population0::Array, pop_size::Int)
+    contestant1 = population0[Int(ceil(rand()*pop_size))] #Choose a random image from the population, using ceil because 0 is not a valid index
+    contestant2 = population0[Int(ceil(rand()*pop_size))]
+
+    while contestant1 == contestant2 #If it happened to pick the same individual twice, change it until you have 2 unique individuals
+        contestant2 = population0[Int(ceil(rand()*pop_size))]
+    end
```
Comment (f6v) [inline src/polygon.jl]: Maybe it'd be a bit cleaner to exclude the first contestant from the array before choosing the second one? You could use `filter` function.

File: src/polygon.jl [hunk 21]
```diff
@@ -0,0 +1,508 @@
...
+
+    if contestant1[3] <= contestant2[3] # Lower score than contestant 2 means he's better
+        winner = contestant1
+    else
+        winner = contestant2
+    end
+    return winner
```
Comment (f6v) [inline src/polygon.jl]: Can it be:
```
return (contestant1[3] <= contestant2[3] ? contestant1 : contestant2)
```
https://stackoverflow.com/a/39790132/468311

File: src/polygon.jl [hunk 22]
```diff
@@ -0,0 +1,508 @@
...
+
+    badTriangle = true
+    p1 = nothing
+    p2 = nothing
+    p3 = nothing
+
+    while badTriangle
+        p1, p2, p3 = round.([triangle.p1, triangle.p2, triangle.p3] + (rand(3) .- 0.5)*(n/10) + (rand(3) .- 0.5)*(m/10)*im)
+        # Position of all 3 points is changed, with the change in x and y value scaling with the width and height of the canvas respectively
+        t = Triangle(p1, p2, p3, triangle.color)
```
Comment (f6v) [inline src/polygon.jl]: I have a hard time following one-letter variable names. Can `t` be more descriptive? I can live with it if it's derived from the mathematical formula, but I don't think it's the case here.

File: src/polygon.jl [hunk 23]
```diff
@@ -0,0 +1,508 @@
...
+    childTriangles = Vector{Triangle}(undef, number_triangles)
+    for i in 1:number_triangles
+        if rand() > 0.5
+            childTriangles[i] = population1[1][i]
+        else
+            childTriangles[i] = population2[1][i]
+        end
+    end
+
+    child = [childTriangles, population1[2], population1[3]]
```
Comment (f6v) [inline src/polygon.jl]: can just write return [childTriangles, population1[2], population1[3]]

File: src/polygon.jl [hunk 24]
```diff
@@ -0,0 +1,508 @@
...
+"""
+function triangleevolution(image::String = "src/figures/TotoroTester.jpeg"; number_triangles::Int = 25, generations::Int = 50, pop_size::Int = 70, elitism_freq::Number = 0.10, newblood_freq::Number =  0.03, mutation_freq::Number = 0.07, gifname = nothing, fps = 2)
+    
+    makegif = !isnothing(gifname) # Only make the result into a gif if the user enters a name for it
+
+    elite_size = Int(round(pop_size*elitism_freq)) # The elite of every generation (best individuals) will get a free pass to the next generation
+    newblood_size = Int(round(pop_size*newblood_freq)) # A part of the population will be replaced with new, random individuals every generation.
+    # This keeps our gene pool fresh and sparkly!
+
+    if image[1:5] == "https" #file from internet: credits to Dan Getz from stackoverflow for this
```
Comment (f6v) [inline src/polygon.jl:496]: I wish people would credit my StackOverflow answers in their codebases :grinning:

**Review by f6v — COMMENTED** [review]
It's interesting project and you did a lot of work!
Please see the comments, I wish there were more descriptive names and some references. Other than that, looks good.

**Comment by kirstvh** [comment]
Hi, 

your project looks really nice!

• Tests did run well

• nice use of structure and abstract type in the code

• Your enthusiasm in the tutorial is infectious, I really liked going through it! There is also a lot of humor in the code of the functions itself, but sometimes it could be more informative to use more standard names/explanations.

• Would your approximation with triangles work better If you use more of them and smaller ones?

• It seems that the ImageMagick package is also a dependency? (notebook)

• Documentation: The inputs and outputs of the functions are not always mentioned/explained, I think this would help to better comprehend everything, e.g. which format should input picture have, what is output of triangleevolution function, what is input and output of STMOZOO.polygon.getboundaries…

• There are no assert commands in the functions to warn and inform user about wrong inputs

• There are some random calculations in generatetriangle that are not explained

• the triangleevolution function in the notebook loads very slowly (already running for more than 5000 seconds, is this normal? I hope it will be loaded by tonight 😉 )

Cheers,
Kirsten

File: Project.toml (hunk 1 above)
Comment (bspanoghe) [reply Project.toml]: That's very strange, I don't even have this package installed. Are you certain?

Comment (f6v) [reply Project.toml]: I'm on macOS, maybe that's an OS-specific dependency.

File: test/polygon.jl (hunk 11 above)
Comment (bspanoghe) [reply test/polygon.jl:106]: I try not to include anything in my code I don't understand, but I'll look into the map function, thanks!

File: src/polygon.jl (hunk 17 above)
Comment (bspanoghe) [reply src/polygon.jl:270]: I think this is the only comment I didn't integrate for a good reason because I'm running out of time. Sorry 😕

File: test/polygon.jl (hunk 11 above)
Comment (f6v) [reply test/polygon.jl:106]: Essentially, `map(_ -> checktriangle(generatetriangle(m, n), m, n), 1:100)` will run `generatetriangle` 100 times and return an array of length 100 with the results of each function call. 
`|> any` will pass that array to `any` function, which will return `true` if at least one of the elements of the array is `true`. 
So `[true, false, false] |> any` returns `true`, and `[false, false, false] |> any` returns `false`. 

In summary, that's a more compact way to write code. But it's up to you. This paradigm is also used in `tidyverse` packages for `R` language. If you ever used `dplyr`, you've probably seen `%>%` which is similar to `|>`.

Comment (bspanoghe) [reply test/polygon.jl:106]: Thank you for the explanation! I'm indeed familiar with %>%, nice to know Julia has a variant of this!
